fs/mmap/fs_munmap.c

```diff
@@ -158,6 +159,32 @@ static int file_munmap_(FAR void *start, size_t length, bool kernel)
   nxsem_post(&g_rammaps.exclsem);
   return ret;
 #else
+
+#ifdef CONFIG_MM_VM_MAP
+  const FAR struct vm_map_entry *map;
+  struct file tmp_file;
+
+  /* Find the first mapping containing the start address */
+
+  map = vm_map_find(start);
+
+  /* If this is a file-backed mapping, call ioctl to do unmap */
+
+  if (map && map->type == VM_MAP_FILE && map->id.inode)
+    {
+      /* Create a temporary file struct and make ioctl */
+
+      tmp_file.f_oflags = 0;
+      tmp_file.f_pos    = 0;
+      tmp_file.f_inode  = map->id.inode;
+      tmp_file.f_priv   = NULL;
```

Comment:
It's very common that driver's open callback will do some initialization and assign file->priv to allocated memory.

Reply:
yes, but munmap is not necessarily done on a file, and there is no file passed to it. Also, the file might not exist at the time of unmap. It is perfectly valid to do e.g. "fd = open("myfile",...); ptr = mmap(..,fd,...); close(fd); munmap(ptr, length);

In the above case there is no file existing. So there is really no good options here; you just have to pass a temporary file structure.

Reply:
> yes, but munmap is not necessarily done on a file, and there is no file passed to it. Also, the file might not exist at the time of unmap. It is perfectly valid to do e.g. "fd = open("myfile",...); ptr = mmap(..,fd,...); close(fd); munmap(ptr, length);
> 

Yes, userspace could close fd before munmap, but kernel could hold file internally if the implementation still need it.

> In the above case there is no file existing. So there is really no good options here; you just have to pass a temporary file structure.
> 

Yes, but at least we should do file_open/file_ioctl/file_close, instead constructing a temp file and forward to file_ioctl.

> And the comment I made above about possibility to make ioctl directly is actually wrong. Of course the driver ioctl expects to get a file pointer as the argument, and it would be quite hackish not to provide one

Reply:
I am afraid that is not possible, there may not be any file at all any more. This is also valid:
"fd = open("myfile",...); ptr = mmap(..,fd,...); close(fd); unlink(fd); munmap(ptr, length);

The reference count keeps the inode still allocated after unlink, but you can't any more open the file since the inode is unlinked from the filesystem (it cannot be found).

And actually, this is even a common pattern in using posix shm, although with posix shm unlink is replaced with shm_unlink. The same thing nevertheless. You need to call the driver with only the information stored in inode; and inode->i_priv is not usable there since it is private for the driver.

Reply:
not right. the struct file is deleted at close(). only the inode is being kept as long as the inode reference counter is > 0

Reply:
I mean you can call file_dup2 in mmap and keep the copy in vm_map_entry.

Reply:
No, why complicate things. This is kernel internal things, and just having a temp file structure wil directly proagate to ioctl, without creating extra structures, opening another file (file_dup2 will call driver's open again) etc. etc. All you want to do here is to call ioctl(MUNMAP).

Reply:
You don't know how driver implement the ioctl, it's wrong to call ioctl with file * which isn't opened before. I haven't saw your shm driver implementation, but if user close fd after mmap and kernel don't hold a reference on it, when do you free your resource?

Reply:
> For the shm, mmap need to increase inode's ref counter and munmap need to decrease that.

So, you increase/decrease inode ref directly without go through vfs layer.

> The allocations are stored on inode's private member (NOT on file pointer, that would be just wrong), and freed when the inode is deleted.

Yes, the real storage is normally associated with inode's private pointer.

> The "file" can be deleted (inode unlinked) before it is closed, it can also be deleted AND closed before the last user munmaps it.

But, the file system or char driver may attach an additional info to file's private pointer in open callback:
https://github.com/apache/incubator-nuttx/blob/master/fs/tmpfs/fs_tmpfs.c#L1461
and reference this pointer in ioctl:
https://github.com/apache/incubator-nuttx/blob/master/fs/tmpfs/fs_tmpfs.c#L1734
Your temp file here will break the implementation like tmpfs, but it is very common practice.

> The deletion of the inode happens when the inodes reference counter goes to 0, and at that point the resources are freed. This happens in inode_release, which is called from the shmfs driver every time you want to de-crease inode reference counter (in munmap). For close and unlink this is called from within vfs.
> 
> There is nothing wrong in calling ioctl with a temporary file *, as long as the file * referes to a proper inode.
> 

Passing file pointer without open break the contract(file_operation_s' callback need happen atfer open callback). It work with your shm driver, but doesn't mean it work with other implementation.
Since mmap is a common layer, we need ensure it work with all possible impleementation.

> Here is a preview for shm: https://github.com/tiiuae/nuttx/pull/40/files
> 
> It is fully working in flat&protected modes, tested in fairly complex environment and with all the possible combinations of shm_open/close/mmap/unmap/shm_unlink

Reply:
Sorry, maybe I didn't manage to explain the thing properly.

It would be wrong to have a file structure passed to munmap, in all cases. Reason for this is that there is no file at all associated with unmap. Even if you have mmapped a file (for mmap there must be a file), it is legal to delete the file before unmapping (close(); unlink()). So any driver relying on file * in *unmap* must be doing something wrong.

In addition, it is not possible to "open" the file (because if it is already unlinked, it doesn't exist, it cannot be searched in inode tree). Also, open and dup both will fail for non-mountable filesystems such as shm etc. If you look at the vfs/fs_open.c, it will never call open from non-mountable fs file_operations.

I could always call call MUNMAP ioctl with null file pointer, but this would require special handling in all the ioctl handlers. which is not nice.

Another option would be, that munmap were a part of file_operations_s, but again, this would touch too many places.

I still think that passing a "dummy" file structure with priv == null is the cleanest solution here.

Reply:
> Sorry, maybe I didn't manage to explain the thing properly.
> 
> It would be wrong to have a file structure passed to munmap, in all cases. Reason for this is that there is no file at all associated with unmap. Even if you have mmapped a file (for mmap there must be a file), it is legal to delete the file before unmapping (close(); unlink()). So any driver relying on file * in _unmap_ must be doing something wrong.
> 
> In addition, it is not possible to "open" the file (because if it is already unlinked, it doesn't exist, it cannot be searched in inode tree).

I don't suggest you open file in unmap, it' wrong place. You need hold the file in mmap with file_dup and file_close in unmap.

> Also, open and dup both will fail for non-mountable filesystems such as shm etc. If you look at the vfs/fs_open.c, it will never call open from non-mountable fs file_operations.
> 

No, please check here:
https://github.com/apache/incubator-nuttx/blob/master/fs/vfs/fs_open.c#L169

> I could always call call MUNMAP ioctl with null file pointer, but this would require special handling in all the ioctl handlers. which is not nice.
> 

You can't do anything if you pass NULL file pointer, if that driver do initialization in xxx_open.

> Another option would be, that munmap were a part of file_operations_s, but again, this would touch too many places.
> 

file_operation_s::ioctl(filep, FIOC_MUNMAP, ...) equals file_operation_s::mumap(filep, ...).  This is just syntactic sugar, no really difference.

> I still think that passing a "dummy" file structure with priv == null is the cleanest solution here.

Reply:
It is just confusing when talking about file_open etc. at unmap. Again, at unmap there is no file related, even though there may be an inode. I am not allocating an extra dummy file structure for each mmap, just because you say so. It is conceptually wrong, complicated and waste of resources. The current implementation doesn't change the operation of existing drivers (such as tmpfs). You don't need an open file, or a linked file, in order to do unmap. So again, I am not forcing one to exist. There is no information that any driver could use from file->priv at unmap, nor can it use flags or anything else but the inode pointer. For other ioctls, what you write is true, but not for unmap.

Reply:
Here is how Linux kernel done:
Hold file reference count and call file_operations::mmap in mmap_region
https://github.com/torvalds/linux/blob/master/mm/mmap.c#L1790-L1791
Release vm_file in unmap.
https://github.com/torvalds/linux/blob/master/mm/mmap.c#L186-L187

Reply:
Yes, linux has a reference counter on the file structure. It doesn't allocate another full file struct on every mmap, but keeps the same kernel side file structure locked. Which makes sense. In NuttX we only have reference counting on inode, which also makes sense. It is different implementation.

imagine:

fd = open(file); /* This creates the kernel side reference, and stores the *file specific* private data in file->priv */

mmap(...,fd,...); /* now assume that this dups the kernel side reference, "file struct", creating another file specific private data in file->priv calling "open" again, and stores that in vm_map */

close(fd); /* This closes the file and deletes the first kernel side file struct */

munmap(ptr...); /* this would now get file->priv from the *second* open call, from dupping in mmap. How is this file private data related to the origina mmap??? */







 NuttX doesn't have it